### DISCUSSION TITLE
Apply extractVersion to datasource but not manager

### DISCUSSION THREAD

**Discussion by samithShetty** [renovatebot/renovate#10665] [Request Help] [answered]
Hi

Apologies for the longer question, most of this is added context for why I am implementing this solution, my actual question is near the end.

I'm working on self-hosted renovate:25 in bitbucket-server. My current use of Renovate is to update references to specific repos ( that are also contained within the same bitbucket server that I am updating dependencies in) using the regex manager and git tags. Our tag versioning scheme is a hybrid of immutable tags that follow semver (but with 'v' prepended, e.g. `v4.8.0`) and mutable tags for each existing major release that resolve to be the latest patch version of that major release (e.g. `v4` = `v4.8.0`, assuming that is the most recent minor/patch release). By default, Renovate properly updates along the immutable tag versions, and I am now trying to make Renovate update to the mutable tags instead, so my desired behavior is:

- Major updates always go to the mutable, e.g. `v3.7.1 -> v4`, `v3 -> v4` (**✓** I have accomplished this)
- Separate PRs for multiple major updates, e.g. `v2 -> v3`,  `v2 -> v4` (**✓** I have accomplished this)
- Outdated minor/patch versions (or if possible, all immutable versions) also get PRs to mutable tag version for that major release, e.g. `v3.6.0 -> v3`   (❌ This is what I need help with)

I have accomplished the first 2 behaviors by using extractVersion in my package rules to essentially take a slice of the immutable that renovate normally updates to, in order to derive the mutable version. Here's the packageRule where I do so:
```js
packageRules: [
    { 
        matchDatasources: ["git-tags"],
        separateMultipleMajor: true,
        extractVersion: "^(?<version>v\\d+)", //Converts immutable tag version from datasource to mutable tag (e.g. v4.8.0 -> v4)
        versioning: "regex:^v(?<major>\\d+)(\\.(?<minor>\\d+))?(\\.(?<patch>\\d+))?"
    }
]
```
Implementing this solution however has seemingly eliminated minor updates however, so I don't get any PRs that simply migrate to the mutable (e.g. `v3.7.0 ->v3`). I believe this is potentially due to the fact that when the versions are read from my package files by my regexManager(see below), they are also converted by this extractVersion statement (this is **not** intended), so Renovate interprets each version as it's corresponding (but not equivalent) mutable tag and therefore doesn't think it needs a minor update. I believe that if I could somehow configure Renovate to interpret the versions found in the package files normally (not applying extractVersion), this issue would likely be fixed

**So, my question is:** Is it possible to apply this regex-slicing package rule only to the git-tags that I am reading as a datasource to update versions, but leave the current version that is read from my package files alone? (or is there a less-hacky alternative solution to what I am trying to accomplish?) 

<details>

<summary> Click me for regexManager + example dependency version entry </summary> 

For reference, this is one of the regex managers I am using, the other operates similarly but for a different dependency:
<p>

```js
    regexManagers: [
        {
            fileMatch: ['^Jenkinsfile$'],
            matchStrings: ["library 'pipeline-utilities@v(?<currentValue>.*?)'"],
            depNameTemplate: "pipeline-utilities",
            datasourceTemplate: "git-tags",
            lookupNameTemplate: "https://bitbucket.int.<company>.com/project/pipeline-utilities.git" //Module Repo URL
        }
    ]
```

and the line from it's respective package file that it grabs the current version from:

```
library 'pipeline-utilities@v3.7.1'
```
</p>
</details>

Thanks, let me know what else I can provide to help

**Comment by viceice** [comment]
I don't understand why you want to switch to mutable tags? if you use renovate, you can let him automerge any minor or patch updates if pipeline succeds.

If you pin to mutable tags, you never get an update by renovate, beside major updates.

**Reply by samithShetty** [reply]
Thanks for the reply.

I'm not sure that autoMerge works in the company bitbucket-server, as there seems to be some sort of automatic approval requirement from Code Owners for every PR, hence why we want to leverage mutable tags to reduce PR clutter and required approvals. 

The second point would work, but we have a lot of app teams using outdated immutable and one of the goals of using Renovate is to get them on mutables (and later major versions) in the first place.

Also, I just thought of a (admittedly hacky) potential solution, but first I wanted to clarify my understanding of packageRules, extractVersion, and regexManagers:
- does packageRules apply to an entire dependency regardless of if I only match it to a datasource (as I did in the example)? So for example, my pipeline-utils module (regexManager provided above) will have all of the rules that I matched to the git-tags datasource (since that's the datasourceTemplate of that regexManager), even when it's just being read from the file and not retrieved through a git-tag? 

- Does extract version (when used in either the regexManager or packageRules) affect the way that the dependency is interpreted when read by the manager, in addition to the datasource? or is it supposed to just be the datasource?

- Does the renovate support regex conditional logic? I am not familiar with regex or the way its logic works, but I think I can use it to get what I want

My hack solution idea is to build conditional logic into the regex of the regexManager/packageRule to make it capture/extract a different group depending on if there are quotations (' or ") after the version, since there always will be in the files (see example above), but never will be in the git-tags. Through this, I could use regex conditional logic to extract a different version from the git-tags datasource version (which should take a slice containing the major) vs what the manager reads (which should just be interpreted normally). I haven't tested it yet and my understanding of the way things work could definitely be thoroughly, thoroughly flawed haha, so let me know if this immediately seems impossible or I'm missing something.
Thanks

**Reply by samithShetty** [reply]
Update: I've been trying to implement my solution. I found a couple of regex patterns that properly matched in https://regex101.com/ and accomplished what I was trying to do, but when put in renovate they all failed. I then realized renovate uses RE2, which seems to be the regex engine that is least capable of doing what I am attempting to do, as it removed many logic features (notably, look aheads), in order to be more efficient, and I'm not quite sure if what I'm trying to do is possible at all with RE2. Is it by chance possible to reconfigure renovate to use a different regex engine? or do any regex wizards know a way to conditionally change what my <version> capture group is grabbing?

**Reply by viceice** [reply]
We use RE2 to not be vulnerable against bad regexes. So no chance to switch.

**Comment by viceice** [comment, marked as answer]
`v3.6.0 -> v3` is not possible, as renovate reads `v3` as `v3.0.0` which is lower than `v3.6.0`.

Please provide a minimal reproduction repo, then we can maybe find a workaround.

**Reply by rarkins** [reply]
Renovate generally tries to *update* versions. Sometimes it *pins* them to more specific versions, and sometimes it rolls them back to an older version if the current one has been yanked from the registry. It doesn't really have the concept of `3.6.3` -> `3` although I think you seem aware of that. It might be possible to hack it in but I can't think of any way.

**Reply by samithShetty** [reply]
> `v3.6.0 -> v3` is not possible, as renovate reads `v3` as `v3.0.0` which is lower than `v3.6.0`.

Ah ok this makes sense now, thank you. 

One final clarification: When I use extractVersion on the tags in the datasource, will it interpret as the original version, then apply extractVersion just for the replace string? If so, I have one more hack idea...

As for the hack, I had one more idea of using handlebars in the extractVerisionTemplate to get the conditionally applied extractVersion rather than trying to do it inside the regex itself. Looks like there's an example in the documentation for getting an if else in the handlebars for the versioningTemplate, I'll try doing that inside extractVersionTemplate and see if that works, otherwise, I'll probably just accept how it is or move back to using immutables.

Thanks for your help!

Update: handlebars hack didn't work, I fundamentally misunderstood how extractVersion works. Ultimately, all I'm trying to do is the strip the version similar to how it's done with extractVersion in the [documentation](https://docs.renovatebot.com/configuration-options/#extractversion), but if it is trying to resolve the version after performing the strip, then yeah I guess it's not possible...
